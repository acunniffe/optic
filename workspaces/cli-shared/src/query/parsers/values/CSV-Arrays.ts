import { IQueryStringValueParser } from '../../query-parser-interfaces';
import { CSVArrayParserOptions } from '@useoptic/cli-config/build/helpers/query-config-interfaces';

export class CSVArrayParser implements IQueryStringValueParser {
  options: CSVArrayParserOptions;
  constructor(options: CSVArrayParserOptions) {
    this.options = options;
  }

  parse(key: string, values: string[]): any {
    const firstValue = values[0];
    return firstValue.split(/,\s*/);
  }

  shouldHandle(key: string, values: string[]): boolean {
    if (
      Array.isArray(this.options.keys) &&
      this.options.keys.includes(key) &&
      values.length === 1
    ) {
      return true;
    } else {
      return false;
    }
  }
}
