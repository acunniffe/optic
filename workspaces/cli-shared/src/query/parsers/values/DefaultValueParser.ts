import { IQueryStringValueParser } from '../../query-parser-interfaces';

class DefaultValueParser implements IQueryStringValueParser {
  parse(key: string, values: string[]): any {
    if (values.length === 1) {
      return values[0];
    } else {
      return values[values.length - 1];
    }
  }

  shouldHandle(key: string, values: string[]): boolean {
    return true;
  }
}

export const DefaultValueParserInstance = new DefaultValueParser();
