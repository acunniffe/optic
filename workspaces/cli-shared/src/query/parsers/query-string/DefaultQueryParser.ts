import {
  IQueryParser,
  IQueryStringKeyValues,
  IQueryStringValueParser,
} from '../../query-parser-interfaces';
import { DefaultValueParserInstance } from '../values/DefaultValueParser';

export class DefaultQueryParser implements IQueryParser {
  valueParsers: IQueryStringValueParser[];

  constructor(parsers: IQueryStringValueParser[]) {
    this.valueParsers = parsers;
  }

  parse(rawQueryString: string): object {
    const values = this.parseToKeysAndValues(rawQueryString);
    const final = {};
    Object.entries(values).map((i) => {
      const [key, values] = i;

      const firstParser = this.valueParsers.find((i) =>
        i.shouldHandle(key, values)
      );
      if (firstParser) {
        // @ts-ignore
        final[key] = firstParser.parse(key, values);
      } else {
        // @ts-ignore
        final[key] = DefaultValueParserInstance.parse(key, values);
      }
    });
    return final;
  }

  parseToKeysAndValues(rawQueryString: string): IQueryStringKeyValues {
    return querystring(rawQueryString);
  }
}

function querystring(query: string): IQueryStringKeyValues {
  let parser = /([^=?&]+)=?([^&]*)/g,
    result = {},
    part;

  const pastKeys: string[] = [];

  while ((part = parser.exec(query))) {
    let key = decode(part[1]),
      value = decode(part[2]);

    if (
      key === null ||
      value === null ||
      (key in result && !pastKeys.includes(key))
    ) {
      console.log(`skipping extra ${key}`);
      continue;
    }

    // @ts-ignore
    if (result[key]) {
      // @ts-ignore
      result[key].push(value);
    } else {
      pastKeys.push(key);
      // @ts-ignore
      result[key] = [value];
    }
  }

  return result;
}

function decode(input: string) {
  try {
    return decodeURIComponent(input.replace(/\+/g, ' '));
  } catch (e) {
    return null;
  }
}
