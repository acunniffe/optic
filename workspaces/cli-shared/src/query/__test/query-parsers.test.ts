import { DefaultQueryParser } from '../parsers/query-string/DefaultQueryParser';
import { CSVArrayParser } from '../parsers/values/CSV-Arrays';

describe('default query parser', () => {
  const parser = new DefaultQueryParser([]);

  test('simple query string', () => {
    const qs = '?foo=bar&bar=foobar';
    expect(parser.parseToKeysAndValues(qs)).toMatchSnapshot();
    expect(parser.parse(qs)).toMatchSnapshot();
  });

  test('same key, multiple times', () => {
    const qs = '?foo=1&foo=2&foo=3';
    expect(parser.parseToKeysAndValues(qs)).toMatchSnapshot();
    expect(parser.parse(qs)).toMatchSnapshot();
  });

  test('empty query string', () => {
    const qs = '';
    expect(parser.parseToKeysAndValues(qs)).toMatchSnapshot();
    expect(parser.parse(qs)).toMatchSnapshot();
  });
});

describe('csv arrays', () => {
  const parser = new CSVArrayParser({
    keys: ['colors'],
  });

  test('parses a csv string to an array', () => {
    expect(
      parser.shouldHandle('colors', ['red,green,yellow'])
    ).toMatchSnapshot();
    expect(parser.parse('colors', ['red,green,yellow'])).toMatchSnapshot();
  });

  test('other key will not be handled', () => {
    expect(parser.shouldHandle('shapes', ['square,circle'])).toMatchSnapshot();
  });

  test('array of one item is still parsed as csv array', () => {
    expect(parser.parse('colors', ['red'])).toMatchSnapshot();
  });
});
