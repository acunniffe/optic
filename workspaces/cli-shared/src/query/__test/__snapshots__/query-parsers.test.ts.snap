// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csv arrays array of one item is still parsed as csv array 1`] = `
Array [
  "red",
]
`;

exports[`csv arrays other key will not be handled 1`] = `false`;

exports[`csv arrays parses a csv string to an array 1`] = `true`;

exports[`csv arrays parses a csv string to an array 2`] = `
Array [
  "red",
  "green",
  "yellow",
]
`;

exports[`default query parser empty query string 1`] = `Object {}`;

exports[`default query parser empty query string 2`] = `Object {}`;

exports[`default query parser same key, multiple times 1`] = `
Object {
  "foo": Array [
    "1",
    "2",
    "3",
  ],
}
`;

exports[`default query parser same key, multiple times 2`] = `
Object {
  "foo": "3",
}
`;

exports[`default query parser simple query string 1`] = `
Object {
  "bar": Array [
    "foobar",
  ],
  "foo": Array [
    "bar",
  ],
}
`;

exports[`default query parser simple query string 2`] = `
Object {
  "bar": "foobar",
  "foo": "bar",
}
`;
